### PR TITLE
Add pattern matches to doc for Option

### PR DESCRIPTION
**Pull request by ashawley**
I've added the pattern match to the doc for many of the `Option` methods.   I've included a suite of tests that I wrote to check my work.

The `Option` data type has been stable in Scala for many years, now.  Its documentation could use getting filled out to be more useful.  It's a commonly used part of the standard library, and it gets a lot of exposure from beginners.

I've built the docs locally and posted them on my Github pages for review:

https://ashawley.github.io/scala/2.12.8/library/scala/Option.html

Here's the current version for comparison:

https://www.scala-lang.org/api/2.12.8/scala/Option.html

PS. I've targeted this change to 2.12.  I figured when 2.12 gets bulk merged on to 2.13, again, I'd work on documenting zip and zip3 -- among other changes in 2.13.

### CODE REVIEW

File: src/library/scala/Option.scala [hunk 1]
```diff
@@ -116,14 +116,38 @@ sealed abstract class Option[+A] extends Product with Serializable {
   self =>
 
   /** Returns true if the option is $none, false otherwise.
+   *
+   * This is equivalent to:
```
Comment (som-snytt) [inline src/library/scala/Option.scala:139]: The edits don't actually respect the surrounding text alignment. https://docs.scala-lang.org/style/scaladoc.html

**Comment by som-snytt** [comment]
This change might be explanatory, but they say avoid pattern match on Option, and this would encourage that instead. https://twitter.com/propensive/status/981805791228387328

**Review by som-snytt — COMMENTED** [review]
I wonder if this sends the best message to beginners. I agree beginners consult this scaladoc, which was the best scaladoc for years.

File: src/library/scala/Option.scala [hunk 2]
```diff
@@ -116,14 +116,38 @@ sealed abstract class Option[+A] extends Product with Serializable {
...
    */
   def isDefined: Boolean = !isEmpty
 
   /** Returns the option's value.
+   *
+   * This is equivalent to:
+   * {{{
+   * option match {
+   *   case Some(x) => x
+   *   case None    => throw new Exception
```
Comment (diesalbla) [inline src/library/scala/Option.scala:167]: Can you make it a `NoSuchElementException` here?

File: src/library/scala/Option.scala [hunk 3]
```diff
@@ -201,12 +272,27 @@ sealed abstract class Option[+A] extends Product with Serializable {
   /** Returns this $option if it is nonempty '''and''' applying the predicate $p to
    * this $option's value returns false. Otherwise, return $none.
    *
+   * This is equivalent to:
+   * {{{
+   * option match {
+   *   case Some(x) if !p(x) => Some(x)
+   *   case _                => None
+   * }
+   * }}}
```
Comment (diesalbla) [inline src/library/scala/Option.scala:300]: You can mention that this is also equivalent to `option.filter( x => ! p(x) )`

**Comment by diesalbla** [comment]
@som-snytt IMHO, style guidelines aside, the goal of an API documentation should be to explain what the result of any method or operation in that API should be. The changes in this PR do help in that goal.

File: src/library/scala/Option.scala [hunk 4]
```diff
@@ -229,6 +315,13 @@ sealed abstract class Option[+A] extends Product with Serializable {
 
   /** Tests whether the option contains a given value as an element.
    *
+   * This is equivalent to:
+   * {{{
+   * option match {
+   *   case Some(elem) => true
```
Comment (NthPortal) [inline src/library/scala/Option.scala]: I believe this is only correct if `elem` is a literal. Otherwise, it either needs to be in backticks? or use a guard (e.g. `case Some(x) if x == elem`)

File: src/library/scala/Option.scala [hunk 5]
```diff
@@ -280,6 +394,13 @@ sealed abstract class Option[+A] extends Product with Serializable {
    * nonempty '''and''' `pf` is defined for that value.
    * Returns $none otherwise.
    *
+   * This is equivalent to:
+   * {{{
+   * option match {
+   *   case Some(x) if pf.isDefinedAt(x) => Some(pf(x))
```
Comment (NthPortal) [inline src/library/scala/Option.scala]: is there a reason not to use `lift` like the actual code does? i.e. `case Some(x) => pf.lift(x)`

File: src/library/scala/Option.scala [hunk 6]
```diff
@@ -313,6 +442,14 @@ sealed abstract class Option[+A] extends Product with Serializable {
...
+   * This is equivalent to:
+   * {{{
+   * option match {
+   *   case Some(x) => List(x)
+   *   case None    => Nil
+   * }
+   * }}}
    */
   def toList: List[A] =
     if (isEmpty) List() else new ::(this.get, Nil)
```
Comment (NthPortal) [inline src/library/scala/Option.scala:467]: not part of this PR, but the fact that it uses `List()` and not `Nil` seems weird.

File: src/library/scala/Option.scala [hunk 7]
```diff
@@ -313,6 +442,14 @@ sealed abstract class Option[+A] extends Product with Serializable {
 
   /** Returns a singleton list containing the $option's value
    * if it is nonempty, or the empty list if the $option is empty.
+   *
+   * This is equivalent to:
+   * {{{
+   * option match {
+   *   case Some(x) => List(x)
```
Comment (NthPortal) [inline src/library/scala/Option.scala:461]: style question: do we generally prefer `List(x)` to `x :: Nil`? I assume this has come up in other places

File: src/library/scala/Option.scala [hunk 8]
```diff
@@ -300,6 +421,14 @@ sealed abstract class Option[+A] extends Product with Serializable {
 
   /** Returns this $option if it is nonempty,
    *  otherwise return the result of evaluating `alternative`.
+   *
+   * This is equivalent to:
+   * {{{
+   * option match {
+   *   case Some(x) => Some(x)
+   *   case None    => Some(alternative)
```
Comment (NthPortal) [inline src/library/scala/Option.scala]: this should be `alternative`, not `Some(alternative)`.

There is also some argument to be made of not unboxing and reboxing `x`, but I'm not sure how important that is. The same applies to the comments for `filter` and `filterNot`.

**Review by NthPortal — DISMISSED** [review]
There are a couple places where the new comments are incorrect (`contains` and `orElse`); the rest are more questions I'm unsure about.

**Comment by jozic** [comment]
What's the purpose of that? 
I can see how people will read it and say: "So should I use pattern match instead of `isEmpty` or instead of `map`?" So it may (not saying it will, but it may) increase confusion, but what's the benefit?

Also tests don't really test the code in docs, `tut` or `sbt-doctest` can fit better here imo

File: src/library/scala/Option.scala (hunk 1 above)
Comment (ashawley) [reply src/library/scala/Option.scala:139]: I'm not sure I follow.  You're talking about what column the "gutter asterisks" are?  I believe I followed the convention throughout the file.

File: src/library/scala/Option.scala (hunk 2 above)
Comment (ashawley) [reply src/library/scala/Option.scala:167]: I'm trying to keep things illustrative, rather than precise.

File: src/library/scala/Option.scala (hunk 3 above)
Comment (ashawley) [reply src/library/scala/Option.scala:300]: Yes, that's also true, but is orthogonal to what I'm proposing.

File: src/library/scala/Option.scala (hunk 2 above)
Comment (som-snytt) [reply src/library/scala/Option.scala:167]: Thanks, I'm going to use that line with my daughter.

Comment (ashawley) [reply src/library/scala/Option.scala:167]: > Thanks, I'm going to use that line with my daughter.

You're welcome.

File: src/library/scala/Option.scala (hunk 4 above)
Comment (ashawley) [reply src/library/scala/Option.scala]: You're right, good catch.

File: src/library/scala/Option.scala (hunk 5 above)
Comment (ashawley) [reply src/library/scala/Option.scala]: I figured a boolean predicate function like `isDefinedAt` was as much as one would want to expose to a beginner about partial functions here, so showing `lift` seems like more cognitive load rather than less.

File: src/library/scala/Option.scala (hunk 6 above)
Comment (ashawley) [reply src/library/scala/Option.scala:467]: Indeed, there are a dozen of them throughout the library.

File: src/library/scala/Option.scala (hunk 7 above)
Comment (ashawley) [reply src/library/scala/Option.scala:461]: There are only a dozen cases of `:: Nil` in the library, but at least twice as many examples of the `List()` constructor.

File: src/library/scala/Option.scala (hunk 8 above)
Comment (ashawley) [reply src/library/scala/Option.scala]: Good catch.  I wasn't consistent with the test, so I tried to fix that as well.

**Comment by ashawley** [comment]
There is one method, `collect`, where showing the equivalent pattern match may not be very illustrative.  This was one of the weak points for the case that the pattern match is an expressive form of what a function does.  This could be because it itself takes a pattern match.  As is, the method's pattern match equivalent that I've written requires exposing the concept of partial functions and assumes more knowledge of the reader.  I spent time thinking about it, but settled on showing the boolean predicate `PartialFunction.isDefinedAt()`, but maybe someone else has a better idea.  

```scala
option match {
  case Some(x) if pf.isDefinedAt(x) => Some(pf(x))
  case _                            => None
}
```

**Comment by ashawley** [comment]
> There is one method, `collect`, where showing the equivalent pattern match may not be very illustrative.

After some thought, maybe I will just remove the pattern match for `collect` since it doesn't seem edifying.  There's an example for using `collect`, and that will need to be sufficient.

**Review by diesalbla — APPROVED** [review]

**Comment by ashawley** [comment]
> This change might be explanatory, but they say avoid pattern match on `Option`, and this would encourage that instead.

Most of the methods are syntactic sugar for the longer pattern match, so presumably people would be encouraged to use the shorter function version rather than the pattern match. The utility of these functions should sell themselves.  The pattern match just helps express what the function does so users use them.

I added the pattern match equivalents "below the fold", and not at the top of the page.  You also have to click on a method and expand its doc to see each one of these.  Hopefully, that avoids over-emphasizing them for you.  The top of the page already has two pattern matches.

I've proposed adding links to the useful functions for `Option` in #7674.

**Comment by SethTisue** [comment]
@ashawley if you squash it, I'll hit "merge"

**Comment by ashawley** [comment]
@SethTisue I've squashed to one commit.
